### ISSUE TITLE
`num_cpus` for ray backend not respected in `run_simulation`

### ISSUE THREAD

**Issue by Svdvoort** [adap/flower#3463]
### Describe the bug

When setting `num_cpus` in `backend_config` for `run_simulation` I would expect this to be parsed to `ray.init` to limit the number of CPUs used for the simulation. However, this argument does not seem to be passed, and instead all available CPUs are used. I often want to leave some CPUs idle for other tasks, and this is currently not possible.

### Steps/Code to Reproduce

import warnings

from typing import Dict

import flwr as fl
import numpy as np

from flwr.client import ClientApp
from flwr.common import NDArrays
from flwr.server import ServerApp
from flwr.server import ServerConfig
from flwr.simulation import run_simulation
from flwr_datasets import FederatedDataset
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import log_loss

N_CLIENTS = 5

class MnistClient(fl.client.NumPyClient):
    def __init__(self):
        partition_id = 1

        # Load the partition data
        fds = FederatedDataset(dataset="mnist", partitioners={"train": N_CLIENTS})

        dataset = fds.load_partition(partition_id, "train").with_format("numpy")
        X, y = dataset["image"].reshape((len(dataset), -1)), dataset["label"]
        # Split the on edge data: 80% train, 20% test
        self.X_train, self.X_test = X[: int(0.8 * len(X))], X[int(0.8 * len(X)) :]
        self.y_train, self.y_test = y[: int(0.8 * len(y))], y[int(0.8 * len(y)) :]

        # Create LogisticRegression Model
        self.model = LogisticRegression(
            penalty="l2",
            max_iter=1,  # local epoch
            warm_start=True,  # prevent refreshing weights when fitting
        )

        # Setting initial parameters, akin to model.compile for keras models
        MnistClient.set_initial_params(self.model)

    def get_parameters(self, config):  # type: ignore
        return self.get_model_parameters(self.model)

    def fit(self, parameters, config):  # type: ignore
        MnistClient.set_model_params(self.model, parameters)
        # Ignore convergence failure due to low local epochs
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            self.model.fit(self.X_train, self.y_train)
        print(f"Training finished for round {config['server_round']}")
        return self.get_model_parameters(self.model), len(self.X_train), {}

    def evaluate(self, parameters, config):  # type: ignore
        MnistClient.set_model_params(self.model, parameters)
        loss = log_loss(self.y_test, self.model.predict_proba(self.X_test))
        accuracy = self.model.score(self.X_test, self.y_test)
        return loss, len(self.X_test), {"accuracy": accuracy}

    @staticmethod
    def get_model_parameters(model: LogisticRegression) -> NDArrays:
        """Returns the parameters of a sklearn LogisticRegression model."""
        if model.fit_intercept:
            params = [
                model.coef_,
                model.intercept_,
            ]
        else:
            params = [
                model.coef_,
            ]
        return params

    @classmethod
    def set_model_params(cls, model: LogisticRegression, params: NDArrays) -> LogisticRegression:
        """Sets the parameters of a sklean LogisticRegression model."""
        model.coef_ = params[0]
        if model.fit_intercept:
            model.intercept_ = params[1]
        return model

    @classmethod
    def set_initial_params(cls, model: LogisticRegression):
        """Sets initial parameters as zeros Required since model params are uninitialized
        until model.fit is called.

        But server asks for initial parameters from clients at launch. Refer to
        sklearn.linear_model.LogisticRegression documentation for more information.
        """
        n_classes = 10  # MNIST has 10 classes
        n_features = 784  # Number of features in dataset
        model.classes_ = np.array([i for i in range(10)])

        model.coef_ = np.zeros((n_classes, n_features))
        if model.fit_intercept:
            model.intercept_ = np.zeros((n_classes,))

def fit_round(server_round: int) -> Dict:
    """Send round number to client."""
    return {"server_round": server_round}

def get_evaluate_fn(model: LogisticRegression):
    """Return an evaluation function for server-side evaluation."""

    # Load test data here to avoid the overhead of doing it in `evaluate` itself
    fds = FederatedDataset(dataset="mnist", partitioners={"train": 10})
    dataset = fds.load_split("test").with_format("numpy")
    X_test, y_test = dataset["image"].reshape((len(dataset), -1)), dataset["label"]

    # The `evaluate` function will be called after every round
    def evaluate(server_round, parameters: fl.common.NDArrays, config):
        # Update model with the latest parameters
        MnistClient.set_model_params(model, parameters)
        loss = log_loss(y_test, model.predict_proba(X_test))
        accuracy = model.score(X_test, y_test)
        return loss, {"accuracy": accuracy}

    return evaluate

def get_client_fn(cid: str):
    return MnistClient().to_client()

# Start Flower server for five rounds of federated learning
if __name__ == "__main__":

    model = LogisticRegression()
    MnistClient.set_initial_params(model)
    strategy = fl.server.strategy.FedAvg(
        min_available_clients=N_CLIENTS,
        min_fit_clients=N_CLIENTS,
        min_evaluate_clients=N_CLIENTS,
        evaluate_fn=get_evaluate_fn(model),
        on_fit_config_fn=fit_round,
    )
    server_config = ServerConfig(num_rounds=2)

    run_simulation(
        server_app=ServerApp(config=server_config, strategy=strategy),
        client_app=ClientApp(client_fn=get_client_fn),
        num_supernodes=250,
        backend_name="ray",
        backend_config={
            "num_cpus": N_CLIENTS,
            "num_gpus": 0,
            "client_resources": {
                "num_cpus": 1,
                "num_gpus": 0,
            },
        },
        verbose_logging=True,
    )

### Expected Results

In this case only 2 actors should be made, and thus only 2 cores/threads should be used. 

### Actual Results

The number of actors that are created is equal to the number of visible threads/cores (for example 16). This is clearly visible in the line: `Constructed ActorPool with: 16 actors`. Monitoring CPU usage (for example using htop) also shows processes on all thread/cores. 

The problem seems to be that none of the `backend_config` options, apart from `client_resources`, are actually parsed to ray.

**Comment by Svdvoort** [comment]
From my own digging, I think the issue lies somewhere in [lines 58-61 in raybackend.py](https://github.com/adap/flower/blob/ca2817562a36449548a22c78780fcfc7d7eff52f/src/py/flwr/server/superlink/fleet/vce/backend/raybackend.py#L58-L61). Here ray init is called, but without any of the parameters that are specified in `backend_config`. 

Later on [in lines 66-75](https://github.com/adap/flower/blob/ca2817562a36449548a22c78780fcfc7d7eff52f/src/py/flwr/server/superlink/fleet/vce/backend/raybackend.py#L66-L75) the client resources and tensorflow settings are separately extracted and used to determine the number of clients. However, this means that the number of cpus explicitly set for ray is completely ignored.

**Comment by jafermarq** [comment]
Hello @Svdvoort, you are right. Currently is not possible to pass arguments directly to `ray.init()` as you point out (making it use of all system resources). 

Le't us know if you'd be keen in opening a small PR to update `RayBackend`. Maybe we can use a `ray_init_args` dict  key inside `backend_config` dict and then pass that to `init_ray()`? what do you think?

**Comment by Svdvoort** [comment]
Yes, I can open a PR to implement this. 

I'm just doubting about the implementation. Currently, `client_resources` is already accepted in the same way that Ray does. Using `ray_init_args`  then seems like an unusual naming convention (compared to not having `ray_client_resources`). I would opt for just `init_args`, to make the naming backend agnostic. In the case above the backend is then just determined by the `backend_name` argument, and the logic for the processing of the `backend_config` needs to be implemented for the different backends. Which I think is okay, since most backends probably will have some notion of init arguments. Let me know what you think and I'll start on a PR.

**Comment by Svdvoort** [comment]
Perhaps at the same time it is logical to move the `init_ray` function to the `RayBackend` class? It seems weird to have it as a "floating", standalone function in the `ray_actor` script in Simulation. It is not specifically related to the simulations right?

**Comment by jafermarq** [comment]
> Yes, I can open a PR to implement this.
> 
> I'm just doubting about the implementation. Currently, `client_resources` is already accepted in the same way that Ray does. Using `ray_init_args` then seems like an unusual naming convention (compared to not having `ray_client_resources`). I would opt for just `init_args`, to make the naming backend agnostic. In the case above the backend is then just determined by the `backend_name` argument, and the logic for the processing of the `backend_config` needs to be implemented for the different backends. Which I think is okay, since most backends probably will have some notion of init arguments. Let me know what you think and I'll start on a PR.

Yes, `init_args` sounds right to me.

**Comment by jafermarq** [comment]
> Perhaps at the same time it is logical to move the `init_ray` function to the `RayBackend` class? It seems weird to have it as a "floating", standalone function in the `ray_actor` script in Simulation. It is not specifically related to the simulations right?

I agree with this. I'll open a PR to move it. Good suggestion.

**Comment by Svdvoort** [comment]
Started a draft PR: #3454 to ensure that double work is not done. Also implements the refactor for `init_ray`

**Comment by Svdvoort** [comment]
Resolved in #3543 / [87a305c](https://github.com/adap/flower/commit/87a305cd68babca4c6c6f0c72798e68a9e52fbf9)
